\documentclass[dvips, a4paper, 11pt,nocolor]{report}

%%%%%%%%%%%%%%%%%%%%%%%% PACKAGES %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

\usepackage{a4wide}
\usepackage[nocolor]{pdfswitch}
\usepackage{texgraphicx}% Pour inclure les pstex_t
\graphicspath{{fig/}{eps/}}
\usepackage[latin1]{inputenc}% pour pouvoir taper des accents directement
\usepackage[T1]{fontenc}% jolie fontes
\usepackage{color}% Pour la couleur (mais pourquoi je l'ai mis ! :
                  % pour xfig)
%\usepackage{colortbl}% Pour faire des tableaux avec des cases colorees
\usepackage{subfigure}% Pour faire des figures en plusieurs parties
\usepackage{boxedminipage}% Pour faire des boxedminipage
\usepackage{xspace}% Pour les espaces Ã  la fin des macros
\usepackage{amssymb} %
\usepackage{times} % Pareil
\usepackage{float} % Pour le positionnement H de figure
\usepackage{vmargin} %
\usepackage{url} %
\usepackage{amsmath}
\usepackage{threeparttable} % pour les notes dans les tables
\usepackage[square]{natbib}
\usepackage{lscape}
\usepackage{placeins}
\usepackage{fancyhdr}
\usepackage{url}
\usepackage{setspace}
\usepackage{alltt}

%\doublespacing

%%%%%%%%%%%%%%%%% COMMANDES %%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\newcommand{\mysinglespace}{\def\baselinestretch{1.1}}
\newcommand{\mydoublespace}{\def\baselinestretch{1.6}}
\newcommand{\treevolve}{\texttt{TREEVOLVE}\xspace}
\newcommand{\phase}{\texttt{phase}\xspace}
\newcommand{\famhap}{\texttt{FamHap}\xspace}
\newcommand{\paup}{\texttt{paup}\xspace}
\newcommand{\phylip}{\texttt{phylip}\xspace}
\newcommand{\paml}{\texttt{paml}\xspace}
%TODO est-ce que phyml s'écrit comme ça ?
\newcommand{\phyml}{\texttt{phyml}\xspace}
\newcommand{\altree}{\texttt{ALTree}\xspace}
\renewcommand{\thetable} {\rm\Roman{table}}
\renewcommand{\thefigure} {\rm\Roman{figure}}
\newcommand{\newchitree}{\texttt{ALTree}\xspace}
\newcommand{\etHTname}{altree-add-S}
\newcommand{\etHT}{\texttt{ALTree-add-S}\xspace}
\newcommand{\rechaploname}{altree-convert}
\newcommand{\rechaplo}{\texttt{ALTree-convert}\xspace}
\newcommand{\acctran}{\texttt{acctran}\xspace}
\newcommand{\deltran}{\texttt{deltran}\xspace}
\newcommand{\chisquare}{chi-square\xspace}
\newcommand{\chisquares}{chi-squares\xspace}
\newcommand{\cmd}[1]{\textit{#1}}
\newcommand{\fn}[1]{\textbf{#1}}
\newcommand{\prog}[1]{\texttt{#1}}
\newcommand{\code}[1]{\texttt{#1}}
\newcommand{\ucode}[1]{\code{\textit{#1}}}
\newcommand{\ccode}[1]{\code{\textbf{#1}}}

\newcommand{\commande}[1]{\texttt{#1}}

\newenvironment{source}{%
\begin{alltt}%
}{%
\end{alltt}%
}

\newenvironment{options}{%
  \bgroup%
  \newcommand{\option}[2]{%
    \code{##1}& ##2 \\%
  }
  \begin{tabular}{ll}
  }{%
  \end{tabular}
  \egroup%
}

%%%%%%%%%%%%%%%%%
\begin{document}
\pagestyle{empty}
\pagestyle{myheadings}
\title{}
\author{Claire Bardel, Vincent Danjean, Pierre Darlu and Emmanuelle Génin}

\maketitle

\tableofcontents

\chapter{Overview of the software}

 This software was designed to perform phylogeny based analysis: first,
it allows the detection of an association between a candidate gene and
a disease, and second, it enables to make hypothesis about the
susceptibility loci.

%\section{Copyright}
It is copyright (c) by Claire Bardel and Vincent Danjean
You are free to distribute this software under the terms of the GNU
General Public License.  The complete text of the GNU General Public
License can be found in the annexe~\ref{GPL} on page~\pageref{GPL}.

It contains three programs: \newchitree, \rechaplo and
\etHT. The connections between these programs are described in Figure~\ref{fig:altree}
%\section{Short description of the programs available in \altree}
\begin{figure}[htbp]
  \includegraphics[width=\linewidth]{overview.fig}\centering
  
  \caption{\altree programs}
  \label{fig:altree}
\end{figure}
\section{\newchitree}
\subsection{Association test}
The test consists in performing series of nested homogeneity tests
(\chisquare) comparing the number of cases and controls in the
different clades defined on the tree. A global p-value is calculated
for the tree by using a permutation procedure such as the one
described by \citet{Ge03} and \citet{Becker04}.

\subsection{Localization of the susceptibility loci}
To perform the localization analysis, for each haplotype $h$, the user
must previously define a new character (called character $S$) whose
state depends on the proportion of cases carrying haplotype $h$ and
optimize it on the haplotype phylogeny.  The program \newchitree then
looks for sites that co-mutates with the character S by calculating a
co-mutation index called $V_{i}$ for each site $i$ and for each
character state transition ($0 \rightarrow 1$ for example). The
higher the $V_{i}$, the higher the probability of $i$ being the
susceptibility site.

\bigskip The method implemented in \newchitree has been fully
described in~\citet{Bardel05}. Please refer to this article for a more
complete description.

\section{\rechaplo}
Before running \newchitree, you will generally have to reconstruct
haplotypes (see section \ref{sec:require} for a description of usable
programs). The output of the haplotype reconstruction programs are
totally different from the input files necessary for the phylogenetic
reconstruction programs. \rechaplo was then written to convert the
outputs of haplotype reconstruction programs to input files for
phylogenetic reconstruction programs. It may be particularly useful if
you want to use \paup because \paup has a very high number of options.
If you use \rechaplo, an input file with all the options necessary to
further run \newchitree is produced.

Currently, \rechaplo can deal with two haplotype reconstruction
programs: \phase~\citep{Stephens01, Stephens03} and
\famhap~\citep{Becker04} and can produce files for three phylogeny
reconstruction programs: \paup~\citep{Swofford02},
\phylip~\citep{Felsenstein04} and \paml~\citep{YangPAML}.

\section{\etHT}\label{description_etiquette}
To perform the localization analysis, a new character $S$ must be
added to each haplotype $h$. The state of $S$ depends on the
proportion of cases carrying the haplotype $h$. You can use your own
criterion to determine the state of $S$ and add it manually to the
input file of the phylogeny reconstruction program that will optimize
the character states changes on the tree.

If you do not want to add the S character manually, you can use \etHT.
The state of the character $S$ is allocated depending on the
proportion ($p_h$) of cases carrying the haplotype $h$ compared to the
proportion $p_0$ of cases in the whole sample.

\begin{itemize}
\item if $p_h < p_0-\epsilon\sqrt{\frac{p_h\times(1-p_h)}{n_h}}$, $S$
  is coded ``C'' or ``0'' (high number of controls);
\item if $p_h > p_0+\epsilon\sqrt{\frac{p_h\times(1-p_h)}{n_h}}$, $S$
  is coded ``G'' or ``1'' (high number of cases);
\item else, $S$ is coded ``?'' (unknown data).
\end{itemize}
with $n_h$ being the number of individuals carrying the haplotype $h$.






\chapter{Installing the software}

The software can run on various linux/Unix platform. %and on MacOS X.

\section{Requirements}
\label{sec:require}

\subsection{Phylogeny reconstruction programs}
Before using \altree, you must build a phylogeny of the haplotypes.
Three phylogeny software are compatible with our program:
\begin{itemize}
\item \paup~\cite{Swofford02}: available at
  \url{http://paup.csit.fsu.edu/}. This software is not free software
  and must be purchased (100\$ for the unix version).
\item \phylip~\cite{Felsenstein04}: freely available at
  \url{http://evolution.genetics.washington.edu/phylip.html} (only
  usable for the association test)
\item \paml~\cite{YangPAML}: freely available at
  \url{http://abacus.gene.ucl.ac.uk/software/paml.html}. As stated by
  its author, \paml is not good at tree making. So we advise you to
  use another software to build the tree (for example,
  \phyml~\cite{Guindon03} and then to use \paml to estimate the
  character states at each node.
\end{itemize}

\paragraph{Note:}
Currently, only parsimony methods implemented in \paup (command
\texttt{set}, option \texttt{criterion} set to ``parsimony'') and in
\phylip (program \texttt{mix}) have been tested. If you want to use
maximum likelihood (ML), we suggest you to use your favorite software
to compute the ML tree and then, to use \paml to estimate the
character states at each node.

\subsection{Required tools}
\prog{perl} is required to run \altree. \prog{perl} version 5.8.7 or
higher should work. Lower version can work, but it has not been
tested.

If you want to build the program from sources, you will also need a
C compiler such as \prog{gcc} and the GNU \prog{make} program. They
are available on most Unix plateforms. %For MacOS X, \prog{gcc} and
%\prog{make} are available with other tools on the development CD of
%the OS.

\section{Installation on a linux platform} 
To install this module type the following:\\

\begin{source}
perl Makefile.PL
make
make test
make install
\end{source}

\noindent If you prefer to install it in your home directory, then
type the following:
\begin{source}
perl Makefile.PL PREFIX=~ 
make
make test
make install
\end{source}

\noindent In this case, do not forget to add \verb+~+\prog{/bin} in your
\prog{PATH} and \verb+~+\prog{/lib/perl/\textit{perl\_version}/} in
\prog{PERL5LIB} if they are not already present. For example:
\begin{source}
PATH=~/bin:$PATH
export PATH
PERL5LIB=~/lib/perl/5.8.7/:$PERL5LIB
export PERL5LIB
\end{source}

\vspace{0.5cm}

%Inclure les differentes platformes sur lesquelles ça tourne.\\
%Voir avec Vince; installation des bibliothèques perl, installation de la bibliothèque C. Avec Linux, paquet debian dispo sur page Vince. Voir si on peut compiler le C pour windows et/ou macOSX.

\chapter{\rechaplo }
This program converts the output of the haplotype reconstruction programs to input for phylogeny reconstruction programs.
\section{Summary of the different options}

\begin{options}
  \option{--version}{program version}
  \option{--short-help|-h}{brief help message}
  \option{--help}{help message with options descriptions}
  \option{--man}{full documentation}
  \option{--first-input-file|-i \ucode{file}}{Input file 1}
  \option{--second-input-file|-j \ucode{file}}{Input file 2 (not mandatory, see
    explanations below)}
  \option{--output-file|-o \ucode{file}}{Output file} 
  \option{--case-control-output|-c  \ucode{file}}{Output containing the
    number of cases/controls}
  \option{--reconstruct-prog|-r \ccode{phase|famhap}}{Name of the haplotype
    reconstruction program}
  \option{--phylo-prog|-p \ccode{paup|phylip}}{Name of the phylogeny reconstruction program}
  \option{--data-type|-t \ccode{DNA|NUM}}{Type of data: DNA (ATGCU) or NUM (0-9)}
\end{options}

\section{Input files}
This program takes as input files the output files of the haplotype recontruction programs. Currently, only \phase (for case/control data) and \famhap (for family data) output files are allowed, but we plan to extend the number of haplotype recontruction programs usable. The haplotype reconstruction program used to generate the input file must be specified after the -r option.

\subsection{Using \phase output file}
Two different cases must be considered:
\begin{itemize} 
\item The case-control status of each individual has been specified in the input file for phase and phase has been run with the -c-1 option. In this case only one input file is necessary for \rechaplo: the phase output file (let's call it out.phase). In this case, the program must be run like this: \\
\begin{source} 
  \rechaploname -r \ccode{phase} -i \ucode{out.phase} -other\_options   
\end{source}
\item The case-control status of each individual has not been specified in the input file for phase. In this case, two input files are necessary: the phase output file (out.phase) and another file which specifies the disease status for each individual (status.phase). This file consists in two rows: the first contains the individual's ID and the second, their disease status (0=control, 1=case). In this case, the program must be run like this: \\
\begin{source}   
  \rechaploname -r \ccode{phase} -i \ucode{out.phase} -j
  \ucode{status.phase}  -other\_options 
\end{source}
\end{itemize}

\subsection{Using \famhap output files}
The program \rechaplo is designed to use files generated with
\famhap15 (and not with \famhap12!), \famhap16 has not been tested
yet. Two options are necessary:
\begin{description}
\item [dp] : to take the disease status into account in the haplotype recconstruction
\item [P] : to make sure that all the haplotypes are present in the output file
\end{description}

Two input files are necessary: the \famhap output file whose name has been chosen by the user (let's call it out.famhap), and the output file called input\_name\_H1\_HAPLOTYPES. In this case, the program must be run like this: \\
\begin{source} 
  \rechaploname -r \ccode{famhap} -i \ucode{out.famhap} -j
  \ucode{H1\_HAPLOTYPES} -other\_options
\end{source}

\section{Output files}
Two kinds of output files can be generated depending on the phylogeny
reconstruction program you want to use (\paml and \phylip use the same
type of input files). The name of the phylogeny reconstruction program should follow the -p option.

\subsection{Generating \paup input files (out.paup)}
The file generated is a nexus file containing the options for \paup
necessary to run \newchitree after \paup. This is only an example of a
paup file: we choose to root the tree using an ancestral sequence, but
this is not necessary and this file should be modified according to
your data. Examples of \paup input files can be found in the test
directory: they are labeled *.paup.  

In the output file, the name of each haplotype is formed by the
concatenation of an haplotype number (Hxxx), and of the number of
cases (mxxx) and controls (cxxx) carrying this haplotype. 
%A character
%is added to each haplotype: its state is 2 or C for all the haplotypes
%and must be set to 1 or G for the ancestral sequence. This character
%is set only to be sure that the phylogeny reconstruction program
%correctly draw the tree. It does not change the results of the
%analysis, and it can be removed from the file.

The output file is not a valid \paup input file.  Some options are
indicated within square brackets and must be specified by the user before
running \paup:
\begin{itemize}
\item the sequence of the ancestral haplotype
\item the maximum number of trees \paup can find
\item the method to optimize character state changes (\acctran/\deltran)
\item the name of the different files generated
\item the number of trees described by paup in the log file (we advise
  you to keep all the trees, and to limit the number of trees that are
  analyzed later, when running \newchitree).
\end{itemize}
The chosen option must be put out of the square brackets because \paup
ignores what is written within square brackets. 



\subsection{Generating \phylip or \paml input files (out.phylip)}
The file generated is the simplest phylip (also used by \paml) format.
The first line contains the number of haplotypes and the number of
sites and the following lines contains an identifier for the haplotype
(Hxxx) and the haplotype sequence. A character is added to each
haplotype: its state is 0 or C for all the haplotypes and must be set
to 1 or G for the ancestral sequence. This character is set only to be sure that the phylogeny reconstruction program correctly draw the tree. It does not change the results of the analysis, and it can be removed from the file. 


In this file, you must either add the sequence of the ancestral haplotype in
the file out.phylip and prepare a file named ``ancestors'' containing this
sequence (don't forget to add the character 1 at the end of the
sequence), or eliminate this sequence and modify the number of
sequences accordingly if you want to use an outgroup to root the tree.

\subsection{The output file correspond.txt}\label{description_correspond} 
This file is automatically generated and the user cannot change it's
name. It consists in lines containing the label of each haplotype and
the number of cases and controls carrying it separeted by spaces or
tabulations. The number of cases carrying a given haplotype is
preceded by the letter ``m'' and the number of controls is preceded
by the letter ``c''.

Example of a correspond.txt file: \\
\begin{tabular}{ccc}
H002 &   m015 &   c001\\
H003 &   m000 &   c001\\
H001 &   m000 &   c002\\
H000\_anc & m000 & c000\\
\end{tabular}

\section{Other options}

\subsection{The -h option: help}
If the program is run with the option -h, a quick help is provided. In this help, the user will find summary of the different options.

\subsection{Type of data}
The user must specify if the data are of type DNA (ATGC) or NUM
(number from 0 to 9, for the current version of the program, numbers
superior to 9 cannot be used). It must be specified after the -r option. 

\subsection{Name of the main output file}
With the -o option, the user chooses the name of the output file. If -o is omitted, the standard output will be used.


 
\chapter{\etHT}
This program adds a new character S to each haplotype corresponding to its disease status.
\section{Summary of the different options}

\begin{options}
  \option{--version}{program version}
  \option{--short-help|-h}{brief help message}
  \option{--help}{help message with options descriptions}
  \option{--man}{full documentation}
  \option{-i \ucode{file}}{Input file 1}
  \option{-j \ccode{correspond.txt}}{Input file 2}
  \option{-o \ucode{file}}{Output file} 
  %-r & Haplotype reconstruction program\\
  \option{-p \ucode{number}}{proportion of cases in the sample}
  \option{-e \ucode{number}}{$\epsilon$ parameter}
  \option{-t \ccode{DNA|SNP}}{data type: SNP or DNA}
  \option{-l}{forces the state of character S to be ``?'' for haplotypes carried by 1 individual}
\end{options}

%\section{Description}
%This program use a paup input file (input.paup) as input file, and generate a new paup input file (input\_et.paup) after having added a new character S to each haplotype. 
%
%S excluded from the reconstruction of the tree, but included for the apomorphy recontrsuction -> modif of the paup block

%les 2 facon de coder le caractere S.

%Pb pour 1 ind.

\section{Input file (-i option)}
The input file must be a valid \paup (nexus) or \paml input file. The
name of the input file must be specified after the -i option

\section{Output file (-o option)}
The output file is a \paup or \paml input file. The S character is coded ``G'' or ``1'' for cases and ``C'' or ``0'' for controls. The name of the output file can be specified after the -o option. If the -o option is not used, the standard output is used. 

In the \paup input file generated, a new option is added, which excludes the character~S from the tree reconstruction process, and includes it in the table of apomorphies. If you want to use \paml, there is no such option. We advise you to reconstruct the phylogeny on the data set without the S-character by using your favorite phylogeny reconstruction program. Then, you give that tree and the data-set with the S-character to \paml to obtain the apomorphie list. 

\section{Other options}

\subsection{The -h option: help}
If the program is run with the option -h, a quick help is provided. In this help, the user will find summary of the different options.

\subsection{Proportion of cases in the sample}
The proportion of cases in the sample must be specified after the -p option

\subsection{The -e option}
Corresponds to the parameter $\epsilon$ (see the description of the program in section \ref{description_etiquette}). If $\epsilon$ is high, haplotypes will more often have a s-character coded ``?''. To give an idea, in our article~\cite{Bardel05}, $\epsilon$ was set to 1.

\subsection{Data type}
The -t option must be followed either by \cmd{SNP} or by \cmd{DNA}. \cmd{SNP}
should be used if you have numerical data (typically SNP data).
 \cmd{DNA} must be used if you have DNA data (A, T, G, C).
Warning: only numbers from 0 to 9 are allowed for microsatellite data.

 \subsection{Haplotypes carried by only 1 individual}
This option -l is not mandatory: if it is present, S is coded ``?'' for all the haplotypes present only once, whatever the disease status of the individual carrying it. If it is not present, the state of S will be chosen according to the formula (see section \ref{description_etiquette}). 

\chapter{\newchitree}
The program can perform either an association test or a localization test. Each option has a long name (which must be preceded by --) and some of them also have a short name (which must be preceded by -).

\section{Summary of the different options}
\begin{options}
\option{--version}{program version}
\option{--short-help|-h}{brief help message}
\option{--help}{help message with options descriptions}
\option{--man}{full documentation}
\option{--association|-a}{perform the association test}
\option{--s-localisation|-l}{perform the localisation using the S character}
\option{--first-input-file|-i \ucode{file}}{result file from phylogeny program}
\option{--second-input-file|-j \ucode{file}}{correspond file (see description below)}
\option{--output-file|-o \ucode{file}}{outfile}
\option{--data-type|-t \ccode{DNA}|\ccode{SNP}}{type of data}
\option{--rootmeth \ccode{outgroup}|\ccode{ancestor}}{rooting method}
\option{--root \ucode{sequence}}{name of the ancestral sequence}
\option{--tree-building-program|-p \ccode{PHYLIP}|\ccode{PAUP}|\ccode{PAML}}{phylip or paup or paml}
\option{--splitmode|-s \ccode{nosplit}|\ccode{chi2split}}{}
\option{--prolongation|-b}{prolongation of branches}
\option{--chi2-threshold|-n \ucode{value}}{threshold value}
\option{--permutations|-r \ucode{number}}{number of permutations to perform}
\option{--trees-to-analyse \ucode{number}}{number of trees to analyse}
\option{--s-site-number \ucode{number}}{position of the $S$ character in the sequence}
\option{--s-site-characters \ucode{anc\_state}->\ucode{der\_state}}{ancestral state -> derived state}
\option{--co-evo|-e \ccode{simple}|\ccode{double}}{simple or double}
\option{--print-tree}{print the tree with the character state changes in the output file}

\end{options}
\section{General options (to be used both for association and for localization)}

\subsection{First input file (option --first-input-file or -i)}
 This file is the output file of the phylogeny reconstruction
 program.  

\subsubsection{If \paup is used}
To run \newchitree, some informations must be present in the input
file. In particular, the apomorphy list and a table containing branch
lenght must appear in the \paup output file. To do so, in the
\cmd{describetrees} command you must use the following options:
\cmd{brlens=yes} and \cmd{apolist=yes};


An example of a \paup input file containig the options necessary to
run \newchitree is provided. % !!! Dans le tar.gz?

\subsubsection{If \phylip is used}
The input file for \newchitree is the output file named ``outfile'' by
\phylip. Currently, \newchitree only works with output data from
\texttt{MIX} (0/1 data). We plan to adapt it to other reconstruction
program soon. To generate a correct input file for \newchitree, you must
use some options depending on your rooting method:
\begin{itemize}
\item If you want to root the tree using an outgroup: in this case, you must root the tree on the chosen outgroup by using the option \cmd{o}.
\item If you want to root the tree using the ancestral character states: you have to prepare an file named \fn{ancestors} containing the ancestral sequence. Then, when running \phylip, you must use the option \cmd{a} (see the \phylip manual for more information).\\
  To run \newchitree, the ancestral sequence must be present in the sample, so we advise you to root the tree using this ancestral sequence as outgroup by using the option \cmd{o}. \\
  In this case, you will have to use the option ``--anc-seq'' when running \newchitree.
\end{itemize}

Moreover, the states at all nodes of the tree must appear in the output file, so you must set the option \cmd{5} to \cmd{yes}.


\subsection{Name of the output file (option --output-file or -o)}
This user can choose the name of the output file by using the --output-file or -o option. If this option is not used, the standard output is used. 

\subsection{Name of the phylogeny program used (option --tree-building-program or  -p )}
After the -p, you should specify which phylogeny reconstruction
program (\paup or \phylip) was used to generate the first input file.

\subsection{Data type  (option --data-type or -t)}
The -t must be followed either by \cmd{SNP} or by \cmd{DNA}. \cmd{SNP}
should be used if you have numerical data (typically SNP data, but
microsatellites are supported by the program in a lesser extend!!! A VOIR).
 \cmd{DNA} must be used if you have DNA data (A, T, G, C).
Warning: only numbers from 0 to 9 are allowed for microsatellite data.
DNA option currently does not work if you have used phylip to
reconstruct the phylogenetic tree.

\subsection{Print tree (option --print-tree)}
With this option, the tree with the character state changes along the
branches will be written in the output file. It may especially be
useful when you are performing the localization analysis, because in this case,  the tree is not written in the output file by default. 


\subsection{How to get help?}

\subsubsection{option --short-help or -h}
This option displays a short help message which recapitulate all the options

\subsubsection{option --help} 
This option displays a message with a description of the different options.

\subsubsection{option --man}
This option displays the man page for the program.

\subsubsection{option --version}
This option gives the number of the version currently used.

\section{Association test (option  --association or -a)}
When the -a option is used, the program will perform the phylogeny based association test. 
\subsection{Options to specify}

\subsubsection{Second input file (--second-input-file or -j option)}
This input file consists in lines contains the label of each haplotype
and the number of cases and controls carrying it separeted by spaces
or tabulations. For a complete description of this file, see section \ref{description_correspond}.

\subsubsection{Root method (--root-method)}
To perform the association test, the tree must be rooted. Depending on your data, you may want to root either on an ancetral sequence (--root-method ancestor) or on an outgroup (--root-method outgroup).

\begin{center}
%\begin{boxedminipage}[width=0.5\linewidth]
\includegraphics[width=0.45\linewidth]{anc_outg.fig}
%\end{boxedminipage}
\end{center}

\subsubsection{Name of the root (--root)}
This option must be followed by the name of the sequence you choose to root the tree (either the outgroup or the ancestor).


\subsubsection{Number of permutations (option --permutations or -r)}
The program can compute a type I error corrected for multiple testing associated with the test by using a double permutation procedure. In this case, the user must define the number of permutations to perform to evaluate the type I error by using the --permutation (or -r) option. This number should be high, but the higher it is, the longer the computation time will be. Depending on the studied data sets, we suggest this number to be chosen between 10000 and 100000.

\subsubsection{Threshold for chi-square significance (option chi2-threshold or -n)}
If you do not want to compute the exact type I error, a significance threshold for the \chisquares can be chosen by the user using the --chi2-threshold (or -n) option. 
 
\subsubsection{Branch prolongation (option --prolongation or -b)}
If the -b option is specified in the command line, the different
branches of the tree will be prolonged. (see
figure~\ref{fig:option_b})

\begin{figure}
\begin{tabular}{|l|c|c|}
  \cline{2-3}
\multicolumn{1}{c|}{}& \includegraphics[width=0.4\linewidth, subfig=1]{option_b.fig} &
\includegraphics[width=0.4\linewidth, subfig=2]{option_b.fig} \\
\cline{2-3}
\multicolumn{1}{c|}{}& Without the -b option & With the -b option \\
\cline{2-3}
\multicolumn{1}{c|}{}&\multicolumn{2}{|c|}{The \chisquares are calculated between\footnote{The
    nosplit pattern of \chisquare is used}:} \\
\hline
  Level 1:& 1 - 2 and 3 & 1 2 and 3 \\
  Level 2:& 2.1 - 2.2 - 3.1 - 3.2 & 1 - 2.1 - 2.2 - 3.1 and 3.2 \\
  Level 3:& 3.2.1 and 3.2.2 & 1 - 2.1 - 2.2 - 3.1 - 3.2.1 and 3.2.2\\
  \hline
\end{tabular}

\caption{Effect of the -b option}
\label{fig:option_b}
\end{figure}

\subsubsection{Ancestral sequence (option --anc-seq)}
This option is only necessary when the tree is rooted using an ancestral sequencce with \phylip (option \cmd{a}). In this case, the ancestral sequence  not being in the output file of \phylip, \newchitree cannot read it directly. So, you have to enter it manually after the --anc-seq option. 

\subsubsection{--splitmode|s nosplit|chi2split}
Option à supprimer car complique beaucoup, non?


\subsection{Description of the output file}
See exemple: \\
The output file show the tree, with the number of cases an controls at each nodes. At the root of the tree, there is a list of the different tests performed on the tree: the level of the test is indicated within square brackets, followed by the number of degrees of freedom is indicated (df=), the value of the \chisquare test and the corresponding p-value. In a second part of the file, a list of the p-values estimated by permutations (but non corrected for multiple testing) for each level of the tree is provided. Then, the last line gives the corrected p-value for the test.


\section{Localization test (option --s-localization or -l)}

\subsection{Options to specify}
\subsubsection{Number of trees (option --trees-to-analyse)}
With this option, you choose the number of trees to
use in the localization test. These trees are randomly sampled without replacement from the input file 1. 

\subsubsection{--s-site-number }
With this option, you specify the position of the S character in the haplotypes. The firt site is numbered 1.  

\subsubsection{ --s-site-characters ancestral state -> derived state}
With this option, you specify which state is the ancestral state and which state is the derived state for the S character. The two states must be separated by the symbol ``->''. For example, if the S character has two states 1 and 2, 1 being the ancestral state, you will use the option as follows: \\
\newchitree [other options ] --s-site-characters 1->2\\
Be careful: this option is case sensitive!

\subsubsection{ --co-evo|e simple|double}
This option allows to choose the way $V_{is}$ are calculated. 

\paragraph{option ''simple''}
This option corresponds to the calculation of $V_{is}$ described in \cite{Bardel05}. Please refer to this publication for more information.


\paragraph{option ''double''} 
This option correspond to a new method to calculate $V_{is}$. This method seems to be more appropriate because it takes into account the two senses of character state changes. \\
Here is a short description of this new calculation method (one  studied tree):
\begin{itemize}
\item Let $E^{0 \rightarrow 1}_{i}$ be the number of expected co-mutations of $S$ (2 character states: T [control] and M [case]) and $i$ (2 character states 0 and 1)
  $$E^{0 \rightarrow 1}_{i}=\frac{(m^{0 \rightarrow 1}_{i}\times
    s^{T \rightarrow M})+(m^{1 \rightarrow 0}_{i}\times
    s^{M \rightarrow T})}{b}$$  \\
  where:
  \begin{description}
  \item[$m^{0 \rightarrow 1}_{i}$~:] nb transitions $0\rightarrow 1$ of $i$
  \item[$s^{T \rightarrow M}$~:] nb transitions $T \rightarrow M$ of $S$  
  \item[$b$~:] nb branches of tree $t$
  \end{description}
\item Let $R^{0 \rightarrow 1}_{i}$ be the number of observed co-mutations of $S$ and $i$ on tree $t$
\item $V^{0 \rightarrow 1}_{i}=0$ is calculated as defined in \cite{Bardel05}:
\end{itemize}

$$\boxed{
  \left\{\begin{array}{ll}
      V^{0 \rightarrow 1}_{i}=0 & if\ E^{0 \rightarrow 1}_{i}=0 \\
      V^{0 \rightarrow 1}_{i}=\frac{\displaystyle{R^{0 \rightarrow 1}_{i}-E^{0 \rightarrow 1}_{i}}}%
      {\displaystyle{\sqrt{E^{0 \rightarrow 1}_{i}}}}
      & if \ E^{0 \rightarrow 1}_{i} \ne 0
    \end{array}
  \right.
}$$

If more than one tree is studied, the $V^{0 \rightarrow 1}_{i}$ must be summed for all the trees.

\subsection{Description of the output file}
The output file contains only a list of the $V_i$ (on ascending order) for the different sites and the different character states transitions. 


\chapter{Tutorials}


\section{Case/Control data, analyzed using \paup}
To build these example file, we simulate a data set from real haplotypes (12 SNPs) from 23 European individual described in the Variation discovery resource project~\cite{SeattleSNPs} (IL13 data). We choose the allele 1 of the first SNP to be the disease susceptibility (DS) allele (frequency: 0.196). Pairs of haplotypes are randomly sampled with replacement to form genotypes and the disease status is obtained by applying the penetrance 0.03 (0 DS allele), 0.06 (1 DS allele) and 0.3 (2 DS allele). This process is carried on until two samples of 100 individuals are obtained. We suppose the phase unknown and we reconstruct it using \phase. Then, we analyze it using all the programs available in this software.

\subsection{Description of the example files}
\paragraph{short description of the different files in the general directory:}
\begin{description}
\item [caco.phase] The file caco.phase is an input file for \phase containing the disease status information for each individual. \phase should be run this way:\\
  \commande{phase -c-1 caco.phase caco.out 100 1 100}
\item [caco.phase.out] This file is the \phase output file obtained with the preceding command line.
\item [caco.phase.out\_*] These file are all the other \phase output file
\item [caco.prepaup] This file is the output file of \rechaplo. It must be modified to become a valid \paup input file
\item [caco.paup] This file is the valid \paup input file
\item [caco.phy] This file is a \phylip format file that can be used by various phylogeny reconstruction program
\end{description}

\paragraph{short description of the different files in the association directory:}
\begin{description}
\item [100\_caco.asso] This file contains the results of the association analysis 
\item  [caco.paup]  This file is the valid \paup input file
\item [correspond.txt] This file contains the number of cases and controls carrying each haplotype (for a more detailed description see section~\ref{description_correspond} 
\item [test.res.log]  This file contains the equipasimonious trees, the table containing the branch length and the apomorphie list. This is the input file for \newchitree
\item [test.tree] This file contains the list of the equiparsimonious trees reconstructed by \paup

\end{description}

\paragraph{short description of the different files in the localization directory:}
\begin{description}
\item [caco.paup] This file is the valid \paup input file
\item [et\_caco.paup] This file is the \paup input file with the new character S, obtained with \etHT 
\item [caco.tree] This file contains the list of the equiparsimonious trees reconstructed by \paup
\item [caco.res.log] This file contains the equipasimonious trees, the table containing the branch length and the apomorphie list. This is the input file for \newchitree
\item [correspond.txt] This file contains the number of cases and controls carrying each haplotype (for a more detailed description see section~\ref{description_correspond} 
\item [caco.loc]  This file contains the results of the localization analysis
\end{description}

The file \fn{caco.phase.out} is a \phase output file containing information on the disease status of each individual.

\subsection{Generating  \paup input file and running \paup}\label{description_paup}
\fn{caco.phase.out} can be converted to an input file for \paup by running this command line:\\
\commande{\rechaplo -i caco.phase.out -t NUM -p paup -r phase -o caco.prepaup}

This command generates the file \fn{correspond.txt} and \fn{caco.prepaup}. The latter is not a valid paup input file: you have to modify it slightly according to your data and your analysis. In this example, we want to root the tree, but we do not know the real ancestral sequence and we do not have an outgroup sequence. So, we choose to root the tree on a consensus ancestral sequence formed by the combination of all the most frequent alleles. \fn{caco.paup} is the resulting paup input file. \paup should then be run like that:\\
\commande{paup caco.paup}

\subsection{Association analysis}
The example files are in the directory /example\_file/case\_control/association/.In this example, we choose to root the tree using the consensus ancestor previously defined and we choose to perform 100 repetitions to estimate the p\_value (100 is too few for a correct estimate of the p\_value, but thus, the computation time is not too long). The program can be run using the following command line:

\commande{\newchitree -i test.res.log -j correspond.txt  -a -t SNP --root H000 --rootmeth ancestor -p paup -b -r 100 > 100\_caco.asso}

In this example, the simulated effect of the disease is very high, so the p\_value is very small and thus, set to 0. As the analyze is only performed on one tree randomly chosen, you will probably not obtain exactly the same output file even if you use the same options.

\subsection{Localization analysis}
\subsubsection{Addition of the character S}
The file \fn{caco.paup} is  used as input file. The proportion of cases haplotypes remaining in the sample is $\frac{105}{181}=0.58$. The program is run using the following command line:\\
\commande{\etHT -i caco.paup -o et\_caco.paup -t SNP -p 0.58 -s 1}

\paup must then be run on the file \fn{et\_caco.paup} (see the directory /examplefiles/case\_control/localization) 

\subsubsection{Localization with \newchitree}
The example files are in the directory /example\_file/case\_control/localization/. In this example, the character S is the site 14, and its ancestral state is 0. We choose to use the option ``double''.

\commande{\newchitree -l -i test.res.log  -t SNP -p paup --trees-to-analyse 89 --s-site-number 14 --s-site-characters "0->1" --co-evo double > caco.loc}

The 2 sites with the highest $V_i$ are site 1 (changing from allele 2 to allele 1)  which is the simulated susceptibility site and site 2 (changing from allele 2 to allele 1)  which is in complete linkage disequilibrium with site 1. 

\section{Trio data, analyzed with  \phylip}
We simulate 100 trios (2 parents and one affected child) genotyped for 10 SNPs. The allele 1 of SNP5 (f=0.2?) is chosen to be the disease susceptibility (DS) allele. The penetrance of the disease is  0.03 (0 DS allele), 0.06 (1 DS allele) and 0.3 (2 DS allele). The haplotypes for all the individuals are reconstructed using \famhap (options dp and P). Two output files: trio.fmh and fam19\_0\_H1\_HAPLOTYPES are obtained

\subsubsection{Generating phylip format file and running \phylip}
The \phylip format file can be obtained by running the following command line:\\
\commande{\rechaplo -i trio.fmh -j fam19\_0\_H1\_HAPLOTYPES -t NUM -p phylip -r famhap -o trio.prephy}

This command generates the file \fn{correspond.txt} and \fn{trio.prephy}. The latter is not a valid phylip format file: you have to modify it slightly according to your data and your analysis. In this example, we choose to root the tree on a consensus ancestral sequence (see~\ref{description_paup}). So, we add this ancestral sequence in \fn{trio.prephy} to form the \fn{trio.phy} file and we build an \fn{ancestors} file, containing this sequence. \phylip is then run on \fn{trio.phy} with the options \cmd{0,37} (tree rooting), \cmd{5} (states at each node are written in the output file) and \cmd{a} (ancestral states must be used). We obtain two files: \fn{outtree} containing all the trees and \fn{outfile} containing the trees and the states at each nodes.  

\subsection{Association analysis}
altree -i outfile  -j correspond.txt  -a -t SNP --root H000\_anc --rootmeth ancestor -p phylip -b -r 100 --anc-seq 11000100010 >100\_trio\_phy.asso



Faire un fichier d'ex pour chaque prog: avoir un fichier \famhap, un fichier \phase et fabriquer les .paup et les .phylip

Puis, montrer un vrai .paup tout complété qui marche. 

avoir aussi les fichiers de sortie de \paup et de \phylip.

Avoir un fichier apres passage par \etHT. 

Avoir fichier de sortie de localisation et d'association

\chapter{URLs where programs can be downloaded}
\section{Haplotype reconstruction programs}
\famhap \url{http://www.uni-bonn.de/%7Eumt70e/becker.html}

\phase \url{http://www.stat.washington.edu/stephens/software.html}

\section{Phylogeny reconstruction program}
\paup  \url{http://paup.csit.fsu.edu/}

\phylip \url{http://evolution.genetics.washington.edu/phylip.html}

\paml  \url{http://abacus.gene.ucl.ac.uk/software/paml.html}

\phyml \url{http://atgc.lirmm.fr/phyml/}
\bibliographystyle{plain}
\bibliography{stage}
\end{document}
\annexe 
\chapter{GNU GENERAL PUBLIC LICENSE}
\label{GPL}
\begin{center}
  {\Large GNU GENERAL PUBLIC LICENSE} \\
  Version 2, June 1991                         
\end{center}
Copyright (C) 1989, 1991 Free Software Foundation, Inc. \\
59 Temple Place, Suite 330, Boston, MA  02111-1307  USA \\
Everyone is permitted to copy and distribute verbatim copies
of this license document, but changing it is not allowed.

\section{Preamble}

  The licenses for most software are designed to take away your
freedom to share and change it.  By contrast, the GNU General Public
License is intended to guarantee your freedom to share and change free
software--to make sure the software is free for all its users.  This
General Public License applies to most of the Free Software
Foundation's software and to any other program whose authors commit to
using it.  (Some other Free Software Foundation software is covered by
the GNU Library General Public License instead.)  You can apply it to
your programs, too.

  When we speak of free software, we are referring to freedom, not
price.  Our General Public Licenses are designed to make sure that you
have the freedom to distribute copies of free software (and charge for
this service if you wish), that you receive source code or can get it
if you want it, that you can change the software or use pieces of it
in new free programs; and that you know you can do these things.

  To protect your rights, we need to make restrictions that forbid
anyone to deny you these rights or to ask you to surrender the rights.
These restrictions translate to certain responsibilities for you if you
distribute copies of the software, or if you modify it.

  For example, if you distribute copies of such a program, whether
gratis or for a fee, you must give the recipients all the rights that
you have.  You must make sure that they, too, receive or can get the
source code.  And you must show them these terms so they know their
rights.

  We protect your rights with two steps: (1) copyright the software, and
(2) offer you this license which gives you legal permission to copy,
distribute and/or modify the software.

  Also, for each author's protection and ours, we want to make certain
that everyone understands that there is no warranty for this free
software.  If the software is modified by someone else and passed on, we
want its recipients to know that what they have is not the original, so
that any problems introduced by others will not reflect on the original
authors' reputations.

  Finally, any free program is threatened constantly by software
patents.  We wish to avoid the danger that redistributors of a free
program will individually obtain patent licenses, in effect making the
program proprietary.  To prevent this, we have made it clear that any
patent must be licensed for everyone's free use or not licensed at all.

  The precise terms and conditions for copying, distribution and
modification follow.

\section{GNU GENERAL PUBLIC LICENSE}
\subsection{TERMS AND CONDITIONS FOR COPYING, DISTRIBUTION AND MODIFICATION}

  0. This License applies to any program or other work which contains
a notice placed by the copyright holder saying it may be distributed
under the terms of this General Public License.  The "Program", below,
refers to any such program or work, and a "work based on the Program"
means either the Program or any derivative work under copyright law:
that is to say, a work containing the Program or a portion of it,
either verbatim or with modifications and/or translated into another
language.  (Hereinafter, translation is included without limitation in
the term "modification".)  Each licensee is addressed as "you".

Activities other than copying, distribution and modification are not
covered by this License; they are outside its scope.  The act of
running the Program is not restricted, and the output from the Program
is covered only if its contents constitute a work based on the
Program (independent of having been made by running the Program).
Whether that is true depends on what the Program does.

  1. You may copy and distribute verbatim copies of the Program's
source code as you receive it, in any medium, provided that you
conspicuously and appropriately publish on each copy an appropriate
copyright notice and disclaimer of warranty; keep intact all the
notices that refer to this License and to the absence of any warranty;
and give any other recipients of the Program a copy of this License
along with the Program.

You may charge a fee for the physical act of transferring a copy, and
you may at your option offer warranty protection in exchange for a fee.

  2. You may modify your copy or copies of the Program or any portion
of it, thus forming a work based on the Program, and copy and
distribute such modifications or work under the terms of Section 1
above, provided that you also meet all of these conditions:

    a) You must cause the modified files to carry prominent notices
    stating that you changed the files and the date of any change.

    b) You must cause any work that you distribute or publish, that in
    whole or in part contains or is derived from the Program or any
    part thereof, to be licensed as a whole at no charge to all third
    parties under the terms of this License.

    c) If the modified program normally reads commands interactively
    when run, you must cause it, when started running for such
    interactive use in the most ordinary way, to print or display an
    announcement including an appropriate copyright notice and a
    notice that there is no warranty (or else, saying that you provide
    a warranty) and that users may redistribute the program under
    these conditions, and telling the user how to view a copy of this
    License.  (Exception: if the Program itself is interactive but
    does not normally print such an announcement, your work based on
    the Program is not required to print an announcement.)

\newpage

These requirements apply to the modified work as a whole.  If
identifiable sections of that work are not derived from the Program,
and can be reasonably considered independent and separate works in
themselves, then this License, and its terms, do not apply to those
sections when you distribute them as separate works.  But when you
distribute the same sections as part of a whole which is a work based
on the Program, the distribution of the whole must be on the terms of
this License, whose permissions for other licensees extend to the
entire whole, and thus to each and every part regardless of who wrote it.

Thus, it is not the intent of this section to claim rights or contest
your rights to work written entirely by you; rather, the intent is to
exercise the right to control the distribution of derivative or
collective works based on the Program.

In addition, mere aggregation of another work not based on the Program
with the Program (or with a work based on the Program) on a volume of
a storage or distribution medium does not bring the other work under
the scope of this License.

  3. You may copy and distribute the Program (or a work based on it,
under Section 2) in object code or executable form under the terms of
Sections 1 and 2 above provided that you also do one of the following:

    a) Accompany it with the complete corresponding machine-readable
    source code, which must be distributed under the terms of Sections
    1 and 2 above on a medium customarily used for software interchange; or,

    b) Accompany it with a written offer, valid for at least three
    years, to give any third party, for a charge no more than your
    cost of physically performing source distribution, a complete
    machine-readable copy of the corresponding source code, to be
    distributed under the terms of Sections 1 and 2 above on a medium
    customarily used for software interchange; or,

    c) Accompany it with the information you received as to the offer
    to distribute corresponding source code.  (This alternative is
    allowed only for noncommercial distribution and only if you
    received the program in object code or executable form with such
    an offer, in accord with Subsection b above.)

The source code for a work means the preferred form of the work for
making modifications to it.  For an executable work, complete source
code means all the source code for all modules it contains, plus any
associated interface definition files, plus the scripts used to
control compilation and installation of the executable.  However, as a
special exception, the source code distributed need not include
anything that is normally distributed (in either source or binary
form) with the major components (compiler, kernel, and so on) of the
operating system on which the executable runs, unless that component
itself accompanies the executable.

If distribution of executable or object code is made by offering
access to copy from a designated place, then offering equivalent
access to copy the source code from the same place counts as
distribution of the source code, even though third parties are not
compelled to copy the source along with the object code.
\newpage
  4. You may not copy, modify, sublicense, or distribute the Program
except as expressly provided under this License.  Any attempt
otherwise to copy, modify, sublicense or distribute the Program is
void, and will automatically terminate your rights under this License.
However, parties who have received copies, or rights, from you under
this License will not have their licenses terminated so long as such
parties remain in full compliance.

  5. You are not required to accept this License, since you have not
signed it.  However, nothing else grants you permission to modify or
distribute the Program or its derivative works.  These actions are
prohibited by law if you do not accept this License.  Therefore, by
modifying or distributing the Program (or any work based on the
Program), you indicate your acceptance of this License to do so, and
all its terms and conditions for copying, distributing or modifying
the Program or works based on it.

  6. Each time you redistribute the Program (or any work based on the
Program), the recipient automatically receives a license from the
original licensor to copy, distribute or modify the Program subject to
these terms and conditions.  You may not impose any further
restrictions on the recipients' exercise of the rights granted herein.
You are not responsible for enforcing compliance by third parties to
this License.

  7. If, as a consequence of a court judgment or allegation of patent
infringement or for any other reason (not limited to patent issues),
conditions are imposed on you (whether by court order, agreement or
otherwise) that contradict the conditions of this License, they do not
excuse you from the conditions of this License.  If you cannot
distribute so as to satisfy simultaneously your obligations under this
License and any other pertinent obligations, then as a consequence you
may not distribute the Program at all.  For example, if a patent
license would not permit royalty-free redistribution of the Program by
all those who receive copies directly or indirectly through you, then
the only way you could satisfy both it and this License would be to
refrain entirely from distribution of the Program.

If any portion of this section is held invalid or unenforceable under
any particular circumstance, the balance of the section is intended to
apply and the section as a whole is intended to apply in other
circumstances.

It is not the purpose of this section to induce you to infringe any
patents or other property right claims or to contest validity of any
such claims; this section has the sole purpose of protecting the
integrity of the free software distribution system, which is
implemented by public license practices.  Many people have made
generous contributions to the wide range of software distributed
through that system in reliance on consistent application of that
system; it is up to the author/donor to decide if he or she is willing
to distribute software through any other system and a licensee cannot
impose that choice.

This section is intended to make thoroughly clear what is believed to
be a consequence of the rest of this License.
\newpage
  8. If the distribution and/or use of the Program is restricted in
certain countries either by patents or by copyrighted interfaces, the
original copyright holder who places the Program under this License
may add an explicit geographical distribution limitation excluding
those countries, so that distribution is permitted only in or among
countries not thus excluded.  In such case, this License incorporates
the limitation as if written in the body of this License.

  9. The Free Software Foundation may publish revised and/or new versions
of the General Public License from time to time.  Such new versions will
be similar in spirit to the present version, but may differ in detail to
address new problems or concerns.

Each version is given a distinguishing version number.  If the Program
specifies a version number of this License which applies to it and "any
later version", you have the option of following the terms and conditions
either of that version or of any later version published by the Free
Software Foundation.  If the Program does not specify a version number of
this License, you may choose any version ever published by the Free Software
Foundation.

  10. If you wish to incorporate parts of the Program into other free
programs whose distribution conditions are different, write to the author
to ask for permission.  For software which is copyrighted by the Free
Software Foundation, write to the Free Software Foundation; we sometimes
make exceptions for this.  Our decision will be guided by the two goals
of preserving the free status of all derivatives of our free software and
of promoting the sharing and reuse of software generally.

\section{NO WARRANTY}

  11. BECAUSE THE PROGRAM IS LICENSED FREE OF CHARGE, THERE IS NO WARRANTY
FOR THE PROGRAM, TO THE EXTENT PERMITTED BY APPLICABLE LAW.  EXCEPT WHEN
OTHERWISE STATED IN WRITING THE COPYRIGHT HOLDERS AND/OR OTHER PARTIES
PROVIDE THE PROGRAM "AS IS" WITHOUT WARRANTY OF ANY KIND, EITHER EXPRESSED
OR IMPLIED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.  THE ENTIRE RISK AS
TO THE QUALITY AND PERFORMANCE OF THE PROGRAM IS WITH YOU.  SHOULD THE
PROGRAM PROVE DEFECTIVE, YOU ASSUME THE COST OF ALL NECESSARY SERVICING,
REPAIR OR CORRECTION.

  12. IN NO EVENT UNLESS REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING
WILL ANY COPYRIGHT HOLDER, OR ANY OTHER PARTY WHO MAY MODIFY AND/OR
REDISTRIBUTE THE PROGRAM AS PERMITTED ABOVE, BE LIABLE TO YOU FOR DAMAGES,
INCLUDING ANY GENERAL, SPECIAL, INCIDENTAL OR CONSEQUENTIAL DAMAGES ARISING
OUT OF THE USE OR INABILITY TO USE THE PROGRAM (INCLUDING BUT NOT LIMITED
TO LOSS OF DATA OR DATA BEING RENDERED INACCURATE OR LOSSES SUSTAINED BY
YOU OR THIRD PARTIES OR A FAILURE OF THE PROGRAM TO OPERATE WITH ANY OTHER
PROGRAMS), EVEN IF SUCH HOLDER OR OTHER PARTY HAS BEEN ADVISED OF THE
POSSIBILITY OF SUCH DAMAGES.

                     END OF TERMS AND CONDITIONS

            How to Apply These Terms to Your New Programs

  If you develop a new program, and you want it to be of the greatest
possible use to the public, the best way to achieve this is to make it
free software which everyone can redistribute and change under these terms.

  To do so, attach the following notices to the program.  It is safest
to attach them to the start of each source file to most effectively
convey the exclusion of warranty; and each file should have at least
the "copyright" line and a pointer to where the full notice is found.

    <one line to give the program's name and a brief idea of what it does.>
    Copyright (C) <year>  <name of author>

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA


Also add information on how to contact you by electronic and paper mail.

If the program is interactive, make it output a short notice like this
when it starts in an interactive mode:

    Gnomovision version 69, Copyright (C) year  name of author
    Gnomovision comes with ABSOLUTELY NO WARRANTY; for details type `show w'.
    This is free software, and you are welcome to redistribute it
    under certain conditions; type `show c' for details.

The hypothetical commands `show w' and `show c' should show the appropriate
parts of the General Public License.  Of course, the commands you use may
be called something other than `show w' and `show c'; they could even be
mouse-clicks or menu items--whatever suits your program.

You should also get your employer (if you work as a programmer) or your
school, if any, to sign a "copyright disclaimer" for the program, if
necessary.  Here is a sample; alter the names:

  Yoyodyne, Inc., hereby disclaims all copyright interest in the program
  `Gnomovision' (which makes passes at compilers) written by James Hacker.

  <signature of Ty Coon>, 1 April 1989
  Ty Coon, President of Vice

This General Public License does not permit incorporating your program into
proprietary programs.  If your program is a subroutine library, you may
consider it more useful to permit linking proprietary applications with the
library.  If this is what you want to do, use the GNU Library General
Public License instead of this License.

\end{document}
